src/labone/nodetree/node.py

```diff
@@ -0,0 +1,1415 @@
+from abc import ABC, abstractmethod
+from contextlib import suppress
+from enum import Enum, IntEnum
+from functools import cached_property
+
+import numpy as np
+from deprecation import deprecated  # type: ignore[import-untyped]
+
+from labone.core.session import KernelSession, ListNodesFlags, NodeType
+from labone.core.session import NodeInfo as NodeInfoType
+from labone.core.value import AnnotatedValue
+from labone.nodetree.helper import (
+    WILDCARD,
+    NestedDict,
+    NormalizedPathSegment,
+    T,
+    UndefinedStructure,
+    build_prefix_dict,
+    get_prefix,
+    join_path,
+    normalize_path_segment,
+    pythonify_path_segment,
+    resolve_wildcards_labone,
+    split_path,
+)
+
+if t.TYPE_CHECKING:
+    from labone.core.helper import LabOneNodePath
+    from labone.core.subscription import DataQueue
+
+
+U = t.TypeVar("U", bound="MetaNode")
+
+
+class TreeProp(t.Generic[T]):
+    """Class for holding information about nodes.
+
+    One such object is meant for holding a specific kind of information which nodes
+    in a tree have. A Node or path can be used to obtain the corresponding
+    information. This way, the information does not have to be stored in the tree
+    structure itself.
+    """
+
+    def __init__(self, path_to_value: dict[LabOneNodePath, T]) -> None:
+        self._path_to_value = path_to_value
+
+    def get(self, node: MetaNode) -> T:
+        """Getting associated information by node."""
+        return self.get_with_path(node.path)
+
+    def get_with_segments(self, path_segments: t.Iterable[NormalizedPathSegment]) -> T:
+        """Getting associated information by collection of path-segments."""
+        return self.get_with_path(join_path(*path_segments))
+
+    def get_with_path(self, path: LabOneNodePath) -> T:
+        """Getting associated information by path."""
+        return self._path_to_value[path]
+
+
+class NodeTreeManager:
+    """Managing relation of one nodetree and one session.
+
+    Executes the initialization procedure of acquiring knowledge about
+    the node-structure. Can be used to find nodes by path, caching them along the way.
+    """
+
+    def __init__(
+        self,
+        session: KernelSession,
+        path_to_info: dict[LabOneNodePath, NodeInfoType],
+    ) -> None:
+        self.session = session
+        self._remembered_nodes: dict[tuple[NormalizedPathSegment, ...], Node] = {}
+        self.path_to_info = path_to_info
+        self._paths: list[LabOneNodePath] = list(self.path_to_info.keys())
+        self._paths_as_segments: list[list[NormalizedPathSegment]] = [
+            split_path(path) for path in self._paths
+        ]
+
+        self._first_layer: NestedDict[
+            list[list[NormalizedPathSegment]] | dict
+        ] = build_prefix_dict(self._paths_as_segments)
+        self.structure_info: TreeProp[NodeInfoType] = TreeProp(self.path_to_info)
+
+    @classmethod
+    async def create(cls, session: KernelSession) -> NodeTreeManager:
```

Comment:
issue: You should not use KernelSession here. We should rather define an interface class that contains all the necessary functions that the node tree requires. This simplifies future work because we can have a node tree based on a fake session easily.

Reply:
😁 maybe I was not clear enough. I was not referring to an abstract class. The KernelSession should not derive from anything. 

I was referring to an interface. In python this is called Protocol (https://peps.python.org/pep-0544/). Its only used for typing and does not have an effect during runtime. In toolkit we use this as well https://github.com/zhinst/zhinst-toolkit/blob/main/src/zhinst/toolkit/nodetree/nodetree.py#L18

The Protocol should be in the node.py class right before `NodeTreeManager`.

Reply:
I know about these two options, which both kind of correspond to an Interface. I chose the abc to have a stricter behavior. Which reasons would you see for the one or the other option?

Reply:
A `Protocol` is typing only, however, an abstract class requires that the passed class derives from it ... So there is a fundamental difference: A `Protocol` just specifies which functions are needed by the code using it, an abstract class states that a class offers specific functionality. In our case, we want to first not the later. 

In addition, the core must not contain any information about the interface the node tree requires ...

So no they are not interchangeable. Change it to `Protocol`

Reply:
`@abstractmethod` is not needed for the interface please remove